task/mgmt-gateway/src/mgs_gimlet.rs

```diff
@@ -3,30 +3,43 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 use crate::{
-    mgs_common::MgsCommon, update_buffer::UpdateBuffer, vlan_id_from_sp_port,
-    Log, MgsMessage, SYS, USART_IRQ,
+    mgs_common::MgsCommon, update::host_flash::HostFlashUpdate,
+    update::sp::SpUpdate, update::ComponentUpdater, vlan_id_from_sp_port, Log,
+    MgsMessage, SYS, USART_IRQ,
 };
 use core::convert::Infallible;
-use core::ops::Range;
 use core::sync::atomic::{AtomicBool, Ordering};
-use drv_gimlet_hf_api::{
-    HfDevSelect, HfError, HfMuxState, HostFlash, PAGE_SIZE_BYTES,
-    SECTOR_SIZE_BYTES,
-};
 use drv_gimlet_seq_api::Sequencer;
 use drv_stm32h7_usart::Usart;
 use gateway_messages::{
     sp_impl::SocketAddrV6, sp_impl::SpHandler, BulkIgnitionState,
-    DiscoverResponse, IgnitionCommand, IgnitionState, PowerState,
-    ResponseError, SpComponent, SpMessage, SpMessageKind, SpPort, SpState,
-    UpdateChunk, UpdateId, UpdatePreparationProgress, UpdatePreparationStatus,
-    UpdatePrepare, UpdateStatus,
+    ComponentUpdatePrepare, DiscoverResponse, IgnitionCommand, IgnitionState,
+    PowerState, ResponseError, SpComponent, SpMessage, SpMessageKind, SpPort,
+    SpState, SpUpdatePrepare, UpdateChunk, UpdateId, UpdateStatus,
 };
 use heapless::Deque;
 use ringbuf::ringbuf_entry_root;
 use task_net_api::{Address, UdpMetadata};
 use userlib::{sys_get_timer, sys_irq_control, UnwrapLite};
 
+// Create type aliases that include our `BufferMutex` size (i.e., the size of
+// the largest update chunk of all the components we update). We could use some
+// kind of const max function over all the components, but we already have
+// static assertions for each component that they fit in this size, so we just
+// manually pick the component with the max (SP updates).
+pub(crate) type BufferMutex = ::update_buffer::BufferMutex<
```

Comment:
I wish this could be an associated type of `UpdateBuffer`, but it looks like inherent associated types are unstable...

Reply:
Oh, the naming actually confused me here: I expected `UpdateBuffer` to be the owner, and `BufferMutex` to be the borrowed thing.

They're just ambiguous enough to go either way; maybe use `UpdateBuffer` for the buffer itself, and `BorrowedUpdateBuffer` for the borrowed thing, to be super explicit?

Reply:
Yeah, I like those names better. :+1:

Reply:
Renamed in 606e9f46